File: Source/Remodel/HeaderScheme.cs
using HmacManagement.Components;

namespace HmacManagement.Remodel;

public class Policy
{
    public Policy(string name, HeaderScheme scheme)
    {
        Name = name;
        Scheme = scheme;
    }

    public readonly string Name;
    public readonly HeaderScheme Scheme;
}

public class HeaderScheme
{
    protected HashSet<Header> Headers = new();

    public HeaderScheme(string name) => Name = name;
    public readonly string Name;

    public IReadOnlyCollection<Header> GetRequiredHeaders() => 
        Headers.ToList().AsReadOnly();

    public void AddRequiredHeader(string name) =>
        AddRequiredHeader(name, name);

    public void AddRequiredHeader(string name, string claimType)
    {
        ArgumentNullException.ThrowIfNullOrEmpty(name, nameof(name));
        ArgumentNullException.ThrowIfNullOrEmpty(claimType, nameof(claimType));

        Headers.Add(new Header(name, claimType));
    }
}

File: src/Scripts/HmacManager/headers/header-scheme.ts
import { IHeaderScheme } from "./i-header-scheme.js";
import { Header } from "./header.js";

export class HeaderScheme implements IHeaderScheme {
    public readonly name: string;
    private readonly collection: Header[] = [];

    constructor(name: string) {
        this.name = name;
    }

    addHeader = (name: string, claimType: string | null = null): void => {
        this.collection.push({ name, claimType: claimType ?? name});
    }

    getHeaders = () => this.collection;
}